numpy/ma/core.py

```diff
@@ -4522,6 +4522,22 @@ def trace(self, offset=0, axis1=0, axis2=1, dtype=None, out=None):
             return D.astype(dtype).filled(0).sum(axis=None, out=out)
     trace.__doc__ = ndarray.trace.__doc__
 
+    def dot(self, other, out=None):
+        am = ~getmaskarray(self)
+        bm = ~getmaskarray(other)
+        if out is None:
+            d = np.dot(filled(self, 0), filled(other, 0))
+            m = ~np.dot(am, bm)
+            return masked_array(d, mask=m)
```

Comment:
Looking at the other ma functions, we want to preserve the input's type - at least one person I found cares, in https://github.com/numpy/numpy/pull/3907. Win insipration fro other funcs, maybe:

    d = np.dot(filled(self, 0), filled(other, 0)).view(type(self))
    d.__setmask__(~np.dot(am, bm))
    return d

The calculation of the mask is cute, btw!